akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -49,62 +51,145 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
+   * tests (and maybe behavior) is missing that checks all states in which WINDOW_UPDATES are received, actually apply WINDOW_UPDATES
+   * revisit entry points into state machines to avoid cycles / unexpected re-entry, between stream handling and multiplexer e.g.
+     * multiplexer.WaitingForNetworkToSendData.onPull => pullNextFrame => outstream.nextFrame => outstream.closeStream => multiplexer.closeStream
+   * mixture of immutable state (explicit states) and mutable components (IncomingStreamBuffer, OutStream)
+
+  */
+
   /**
-   * https://http2.github.io/http2-spec/#StreamStates
+   * States roughly correspond to states as given in https://http2.github.io/http2-spec/#StreamStates.
+   *
+   * Substates were introduced to also track the state of our user-side
+   *
+   * States:
+   *   * Idle                                     | server | client <- Idle, initial state, usually not tracked explicitly
+   *   * OpenReceivingDataFirst                   | server          <- Open, got (request) HEADERS but didn't send (response) HEADERS yet
+   *   * OpenSendingData                          |          client <- Open, sent (request) HEADERS but didn't receive (response) HEADERS yet
+   *   * Open                                     | server | client <- Open, bidirectional, both sides sent HEADERS
+   *   * HalfClosedLocalWaitingForPeerStream      |          client <- HalfClosedLocal, our stream side done, waiting for peer HEADERS
+   *   * HalfClosedLocal                          | server | client <- HalfClosedLocal, all HEADERS sent, just receiving data from peer
+   *   * HalfClosedRemoteWaitingForOutgoingStream | server          <- HalfClosedRemote, waiting for our HEADERS from user
+   *   * HalfClosedRemoteSendingData              | server | client <- HalfClosedRemote, sent our HEADERS, now sending out data
+   *   * Closed                                   | server | client <- Closed, final state, not tracked explicitly
+   *
+   * Server states:
+   *   * Idle -> OpenReceivingDataFirst: on receiving request HEADERS with endStream = false
+   *   * Idle -> HalfClosedRemoteWaitingForOutgoingStream: on receiving HEADERS with endStream = true
+   *   * OpenReceivingDataFirst -> HalfClosedRemoteWaitingForOutgoingStream: after receiving endStream
+   *   * OpenReceivingDataFirst -> Open: after user provided response before request was fully streamed in
+   *   * HalfClosedRemoteWaitingForOutgoingStream -> HalfClosedRemoteSendingData: we sent response HEADERS with endStream = false
+   *   * HalfClosedRemoteWaitingForOutgoingStream -> Closed: we sent response HEADERS with endStream = true
+   *   * HalfClosedRemoteSendingData -> Closed: we sent DATA with endStream = true
+   *   * Open -> HalfClosedRemoteSendingData: on receiving request data with endStream = true
+   *   * Open -> HalfClosedLocal: on receiving response data with endStream = true before request has been fully received (uncommon)
```

Comment:
This transition is quite interesting. HTTP/2 explicitly allows bidirectional streams and also half-closed states at the end of the exchange. Especially, this weird case (also for client, see below), where the response has been sent over completely before the request has been fully received. Even that case is explicitly covered in https://httpwg.org/specs/rfc7540.html#rfc.section.8.1:

> A server can send a complete response prior to the client sending an entire request if the response does not depend on any portion of the request that has not been sent and received. When this is true, a server MAY request that the client abort transmission of a request without error by sending a RST_STREAM with an error code of NO_ERROR after sending a complete response (i.e., a frame with the END_STREAM flag). Clients MUST NOT discard responses as a result of receiving such a RST_STREAM, though clients can always discard responses at their discretion for other reasons.

Do you know if that is actively used in gRPC, @raboof?

Reply:
I'm not sure if that's actively mentioned anywhere, no.